src/main.rs

```diff
@@ -157,3 +265,49 @@ fn parse_line(line: &str) -> anyhow::Result<Option<DVec3>> {
         _ => Ok(None),
     }
 }
+
+fn histogram(
+    positions: &Vec<DVec3>,
+    r_variable: fn(&DVec3) -> f64,
+    (avg, std_dev): (DVec3, DVec3),
+) -> (Vec<(i32, DVec3)>, Vec<(f64, f64)>) {
+    let mut position_set: Vec<DVec3> = positions.clone();
+
+    let cutoff: i32 = 3; // measured in standard deviations
+    let div: i32 = 6;
+    let mut range = (-(cutoff * div)..(cutoff * div))
+        .into_iter()
+        .map(|i| (i as f64) / (div as f64) * r_variable(&std_dev) + r_variable(&avg))
+        .enumerate()
+        .peekable();
+    position_set.sort_by(|a, b| {
+        r_variable(a)
+            .partial_cmp(&r_variable(b))
+            .expect("Histogram: Incomparable values")
+    });
+
+    let division_values = range
+        .clone()
+        .map(|(_, x)| x)
+        .zip(range.clone().map(|(_, x)| x).skip(1))
+        .collect::<Vec<_>>();
+
+    let mut histogram: Vec<(i32, DVec3)> = Vec::new();
+
+    for pos in position_set {
+        while let Some((idx, val)) = range.peek() {
+            if r_variable(&pos) < *val {
+                histogram.push((*idx as i32, pos));
+                break;
+            } else {
+                range.next();
+            }
+        }
+    }
+
+    let histogram = histogram
+        .into_iter()
+        .filter(|(i, _)| *i != 0) //  Removes lower bound of data (atypical data)
+        .collect::<Vec<(i32, DVec3)>>();
```

Comment:
No, this creates an unnecessary allocation. Just remove the element that matches i == 0. Use `retain`.